Add config.yaml definition (#857)
kyma-project/cli: docs/03-04-function-configuration-file.md

          accessStrategies: []
        - path: /path1/something1
          methods:
            - PUT
            - PATCH
            - DELETE
          accessStrategies:
            - handler: noop
        - path: /path1/something2
          methods:
            - GET
            - HEAD
          accessStrategies:
            - config:
                requiredScope:
                    - asd1.asd2.asd3
                    - scope1
                    - scope2
              handler: oauth2_introspection
        - path: /path2
          methods:
            - DELETE
          accessStrategies:
              handler: jwt
            - config:
                jwksUrls:
                    - http://dex-service.kyma-system.svc.cluster.local:5556/keys
                    - http://dex-service.kyma-system.svc.cluster.local:5556
                trustedIssuers:
                    - https://dex.34.90.136.181.xip.io
                    - https://dex.34.90.136.181.xip.io
env:
    - name: REDIS_PASS
      value: YgJUg8z6eA
    - name: REDIS_PORT
      value: "6379"
    - name: REDIS_HOST
      value: hb-redis-enterp-6541066a-edbc-422f-8bef-fafca0befea8-redis.testme.svc.cluster.local
```

> ## Parameters

Reviewer: I found a few missed parameters:
- labels
- source.sourceHandlerName
- source.depsHandlerName
- source.url
- source.repository
- source.reference
- source.baseDir
- source.credentialsSecretName
- resource.limits
- resource.requests
- apiRules.rules.port
- apiRules.rules.path.accessStrategies.config.jwksUrls
- apiRules.rules.path.accessStrategies.config.trustedIssuers
- apiRules.rules.path.accessStrategies.config.requiredScope
- env.valueFrom
- env.valueFrom.configMapKeyRef
- env.valueFrom.configMapKeyRef.Name
- env.valueFrom.configMapKeyRef.Key
- env.valueFrom.secretKeyRef
- env.valueFrom.secretKeyRef.Name
- env.valueFrom.secretKeyRef.Key

See all parameter descriptions:

| Parameter         | Required | Related custom resource | Description                                   |

> | ---------------------------------------- | :------------: | ---------| ---------|

Reviewer: Some values in the config.json, during the applying process, are defaulted by the Cli. Maybe should we add another section or column that can help users to understand what will be applied to the cluster if they missed some non-required fields? For example when the value of the `apiRules.service.port` field is empty then Cli will create apiRule with `80` as the value of the service port

| **source.sourcePath**             | Yes | Function | Specifies the absolute path to the directory with the Function's source code.         |
| **subscriptions.name**           |  Yes | Subscription | Specifies the name of the Subscription custom resource. It takes the name from the Function unless you specify otherwise.    |
| **subscriptions.protocol**           | Yes  | Subscription | Defines the rules and formats applied for exchanging messages between the components of a given messaging system. Subscriptions in Kyma CLI use the [NATS](https://docs.nats.io/) messaging protocol by default. Must be set to `""`.         |
| **subscriptions.filter.dialect**            | No | Subscription | Indicates the filter expression language supported by an event producer. Subscriptions specifying the **filter** property must specify the dialect as well. All other properties are dependent on the dialect being used. In the current implementation, this field is treated as a constant which is blank.    |
| **subscriptions.filter.filters.eventSource**            | Yes | Subscription | The origin from which events are published.         |

> | **subscriptions.filter.filters.eventSource.property**            | Yes | Subscription | Must be set to `source`.         |

Reviewer: If we are documenting objects like the `subscriptions.filter.filters.eventSource.property` or `source` then maybe we can document tables like the `subscriptions.filter.filters`, `subscriptions` or `apiRules`?

Author: I will add thise as well :)

---
title: Function's configuration file
type: Details
---

When you initialize a Function (`init`), CLI creates the `config.yaml` file in your workspace folder. This file contains the whole Function's configuration and specification not only for the Function custom resource but also any other related resources you create for it, such as Subscriptions and API Rules.

## Specification for an inline Function

See the sample `config.yaml` for an inline Function for which code and dependencies are stored in the Function custom resource (CR) under the **spec.source** and **spec.deps** fields. This specification also contains the definition of a sample Subscription and API Rules for the Function:

```yaml
name: function-practical-filip5
namespace: testme

> runtime: nodejs12

Reviewer 2: Now, nodejs14 is default so can we change this runtime?

    - valueFrom:
        configMapKeyRef:
          Name: configmap1
          Key: token-field
    - valueFrom:
        secretKeyRef:
          Name: secret1
          Key: token-field
```

## Specification for a Git Function

See the sample `config.yaml` for a [Git Function](https://kyma-project.io/docs/main/components/serverless/#details-git-source-type) for which code and dependencies are stored in a selected Git repository.

```yaml
name: function-practical-marcin
namespace: iteration-review
runtime: nodejs12
source:
    sourceType: git
    url: https://github.com/username/public-gitops.git
    repository: my-repo
    reference: master
    baseDir: /
    credentialsSecretName: secret2
```

## Parameters

See all parameter descriptions:

| Parameter         | Required | Related custom resource | Default value  | Description|
| ---------------------------------------- | :------------: | ---------| ---------| ------- |
| **name**              | Yes | Function | | Specifies the name of your Function.         |
| **namespace**             | No | Function | `default` | Defines the Namespace in which the Function is created.        |

> | **runtime**             | Yes | Function | | Specifies the execution environment for your Function. The available values are `nodejs12`, `nodejs14`, and `python38`.      |

Reviewer 2: What do you think that add information `nodejs14 is default`?

Reviewer: but it does not default 😃  It's generated by default but it's not defaulted to `nodejs14`

| **subscriptions.protocol**           | Yes  | Subscription | | Defines the rules and formats applied for exchanging messages between the components of a given messaging system. Subscriptions in Kyma CLI use the [NATS](https://docs.nats.io/) messaging protocol by default. Must be set to `""`.         |
| **subscriptions.filter** | No | Subscription | | xxx |
| **subscriptions.filter.dialect**            | No | Subscription | | Indicates the filter expression language supported by an event producer. Subscriptions specifying the **filter** property must specify the dialect as well. All other properties are dependent on the dialect being used. In the current implementation, this field is treated as a constant which is blank.    |
| **subscriptions.filter.filters**            | Yes | Subscription | | Specifies the filtering parameters for the given event.   |
| **subscriptions.filter.filters.eventSource**            | Yes | Subscription | | Defines the origin from which events are published.   |
| **subscriptions.filter.filters.eventSource.property**            | Yes | Subscription | | Must be set to `source`.   |

> | **subscriptions.filter.filters.eventSource.type**           | No  | Subscription | | Must be set to `exact`.    |

Reviewer: ```suggestion
| **subscriptions.filter.filters.eventSource.type**           | Yes  | Subscription | | Must be set to `exact`.    |
```

| **apiRules.name**            | Yes | API Rule | | Specifies the name of the exposed Service. It takes the name from the Function unless you specify otherwise.        |
| **apiRules.gateway**            | Yes | API Rule | `kyma-gateway.kyma-system.svc.cluster.local` | Specifies the [Istio Gateway](https://istio.io/latest/docs/reference/config/networking/gateway/).   |
| **apiRules.service**            | Yes | API Rule | | Specifies the name of the exposed Service.     |
| **apiRules.service.host**            | Yes | API Rule | | Specifies the Service's communication address for inbound external traffic.         |
| **apiRules.rules**            | Yes | API Rule | | Specifies the array of [Oathkeeper](https://www.ory.sh/oathkeeper/) access rules.         |
| **apiRules.rules.methods**            | No | API Rule | | Specifies the list of HTTP request methods available for **apiRules.rules.path** .        |
| **apiRules.rules.accessStrategies**            | Yes | API Rule | | Specifies the array of [Oathkeeper authenticators](https://www.ory.sh/oathkeeper/docs/pipeline/authn/). The supported authenticators are `oauth2_introspection`, `jwt`, `noop`, and `allow`.         |

> | **apiRules.rules.port**            | Yes | API Rule | `80`. | Defines the port on which the Function's Service is exposed. This value cannot be modified. |

Reviewer: ```suggestion
| **apiRules.service.port**            | No | API Rule | `80`. | Defines the port on which the Function's Service is exposed. This value cannot be modified. |
```

> | **apiRules.rules.path**            | Yes | API Rule | `/.*` | Specifies the path to the exposed Service.         |

Reviewer: ```suggestion
| **apiRules.rules.path**            | No | API Rule | `/.*` | Specifies the path to the exposed Service.         |
```

---

> title: Function's configuration file

Reviewer 3: Personally, I think `Function configuration file` sounds better 🤔 But this is also correct, so feel free to leave it.

type: Details
---

> When you initialize a Function (`init`), CLI creates the `config.yaml` file in your workspace folder. This file contains the whole Function's configuration and specification not only for the Function custom resource but also any other related resources you create for it, such as Subscriptions and API Rules.

Reviewer 3: ```suggestion
When you initialize a Function (with the `init` command), CLI creates the `config.yaml` file in your workspace folder. This file contains the whole Function's configuration and specification not only for the Function custom resource but also any other related resources you create for it, such as Subscriptions and API Rules.
```

If the user didn't come directly from the other doc where the `init` command is described, maybe they wouldn't know what exactly is being referred to? But again, this is just a suggestion.

Author: You're right, thx :)

            - DELETE
          accessStrategies:
            - handler: noop
        - path: /path1/something2
          methods:
            - GET
          accessStrategies:
            - config:
                required_scope: ["read"]
              handler: oauth2_introspection
        - path: /path2
          methods:
            - DELETE
          accessStrategies:
              handler: jwt
            - config:
                jwksUrls:
                    - http://dex-service.kyma-system.svc.cluster.local:5556/keys
                    - http://dex-service.kyma-system.svc.cluster.local:5556
                trustedIssuers:
                    - https://dex.34.90.136.181.xip.io
                    - https://dex.34.90.136.181.xip.io
env:
    - name: REDIS_PASS
      value: YgJUg8z6eA
    - name: REDIS_PORT
      value: "6379"
    - name: REDIS_HOST
      value: hb-redis-enterp-6541066a-edbc-422f-8bef-fafca0befea8-redis.testme.svc.cluster.local
    - valueFrom:
        configMapKeyRef:
          Name: configmap1
          Key: token-field
    - valueFrom:
        secretKeyRef:
          Name: secret1
          Key: token-field
```

## Specification for a Git Function

> See the sample `config.yaml` for a [Git Function](https://kyma-project.io/docs/main/components/serverless/#details-git-source-type) for which code and dependencies are stored in a selected Git repository.

Reviewer 3: ```suggestion
See the sample `config.yaml` for a [Git Function](https://kyma-project.io/docs/main/components/serverless/#details-git-source-type) for which code and dependencies are stored in a selected Git repository: 
```

The first `.yaml` description ended with a `:` so it would be nice to add it here as well for consistency.

source:
    sourceType: git
    url: https://github.com/username/public-gitops.git
    repository: my-repo
    reference: master
    baseDir: /
    credentialsSecretName: secret2
```

## Parameters

See all parameter descriptions.

> **NOTE:** The **Default value** column specifies the values that Kyma CLI sets when applying resources on a cluster, if no other values are provided.

| Parameter         | Required | Related custom resource | Default value  | Description|
| ---------------------------------------- | :------------: | ---------| ---------| ------- |
| **name**              | Yes | Function | | Specifies the name of your Function.         |
| **namespace**             | No | Function | `default` | Defines the Namespace in which the Function is created.        |
| **runtime**             | Yes | Function | | Specifies the execution environment for your Function. The available values are `nodejs12`, `nodejs14`, and `python38`.      |
| **labels**             | No | Function | | Specifies the Function's Pod labels. |
| **source**            | Yes | Function | | Provides details on the type and location of your Function's source code and dependencies.         |

> | **source.sourceType**            | Yes | Function | | Defines that you use either inline code or Git repository as the source of the Function's code and dependencies. It must be set either to `inline` or `git`.         |

Reviewer 3: ```suggestion
| **source.sourceType**            | Yes | Function | | Defines whether you use either inline code or a Git repository as the source of the Function's code and dependencies. It must be set either to `inline` or `git`.         |
```

| **source.sourceHandlerName**             | No | Function | `handler.js` (Node.js) or `handler.py` (Python) | Defines the path to the file with your Function's code. Specify it if you want to store source code separately from the `config.yaml`.  This path is a relative path to the one provided in **source.sourcePath**. |
| **source.depsHandlerName**             | No | Function | `package.json` (Node.js) or `requirements.txt` (Python) | Defines the path to the file with your Function's dependencies. Specify it if you want to store dependencies separately from the `config.yaml`. This path is a relative path to the one provided in **source.sourcePath**. |
| **source.url**             | No | Function | | Provides the address to the Git repository with the Function's code and dependencies. Depending on whether the repository is public or private and what authentication method is used to access it, the URL must start with the `http(s)`, `git`, or `ssh` prefix, and end with the `.git` suffix.  |
| **source.repository**             | No | Function | Function name | Specifies the name of the Git repository.  |

> | **source.reference**             | No | Function | | Specifies either the branch name or the commit revision from which the Function Controller automatically fetches the changes in Function's code and dependencies.  |

Reviewer 3: ```suggestion
| **source.reference**             | No | Function | | Specifies either the branch name or the commit revision from which the Function Controller automatically fetches the changes in the Function's code and dependencies.  |
```

| **resources**             | No | Function | | Defines CPU and memory available for the Function's Pod to use.  |
| **resources.limits**             | No | Function | | Defines the maximum available CPU and memory values for the Function.  |
| **resources.limits.cpu**             | No | Function | `100m` | Defines the maximum available CPU value for the Function.  |
| **resources.limits.memory**             | No | Function | `128Mi` | Defines the maximum available memory value for the Function.  |
| **resources.requests**              | No | Function | | Defines the minimum requested CPU and memory values for a Function.  |
| **resources.requests.cpu**              | No | Function | `50m` | Defines the minimum requested CPU value for the Function.  |
| **resources.requests.memory**              | No | Function | `64Mi` | Defines the minimum requested memory value for the Function.  |

> | **subscriptions**   | No | Subscription | | Defines a Subscriptions by which the Function gets triggered to perform a business logic defined in the Function's source code.  |

Reviewer 3: ```suggestion
| **subscriptions**   | No | Subscription | | Defines a Subscription by which the Function gets triggered to perform a business logic defined in the Function's source code.  |
```

| **subscriptions.name**           |  Yes | Subscription | Function name | Specifies the name of the Subscription custom resource. It takes the name from the Function unless you specify otherwise.    |
| **subscriptions.protocol**           | Yes  | Subscription | | Defines the rules and formats applied for exchanging messages between the components of a given messaging system. Subscriptions in Kyma CLI use the [NATS](https://docs.nats.io/) messaging protocol by default. Must be set to `""`.         |

> | **subscriptions.filter** | No | Subscription | | xxx |

Reviewer 3: ```suggestion
| **subscriptions.filter** | No | Subscription | | Defines the filter for the subscriptions. |
```

I'm not sure what the description should be, perhaps the `xxx` means it will be added later? Ignore this comment if so.

Author: Sorry..I missed that...thx a lot for spotting it!
